File: encode-and-decode-tinyurl/encode-and-decode-tinyurl.java
public class Codec {
    Map<String, String> mp = new HashMap<>();
    // Encodes a URL to a shortened URL.
    public String encode(String longUrl) {
        String hashCode=longUrl.hashCode()+""; 
        mp.put(hashCode,longUrl);
        return hashCode;
    }

    // Decodes a shortened URL to its original URL.
    public String decode(String shortUrl) {
        return mp.get(shortUrl);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));

File: 535-encode-and-decode-tinyurl/encode-and-decode-tinyurl.js
const encodeMap = new Map();
const decodeMap = new Map();
let counter = 1;
const baseUrl = 'http://tinyurl.com/';

var encode = function (longUrl) {
    if (encodeMap.has(longUrl)) {
        return baseUrl + encodeMap.get(longUrl);
    }

    const key = counter++;
    encodeMap.set(longUrl, key);
    decodeMap.set(key, longUrl); // Store key to longUrl mapping in decodeMap
    return baseUrl + key;
};


var decode = function (shortUrl) {
    const key = +shortUrl.replace(baseUrl, ''); // Parse key from shortUrl
    return decodeMap.get(key);
};

/**
 * Your functions will be called as such:
 * decode(encode(url));
 */